CRM/MembershipExtras/Job/OfflineAutoRenewal.php

```diff
@@ -0,0 +1,385 @@
+      'non_deductible_amount' => $this->lastContribution['non_deductible_amount'],
+      'fee_amount' => $this->lastContribution['fee_amount'],
+      'net_amount' => $this->lastContribution['net_amount'],
+      'contribution_recur_id'=> $this->currentRecurContributionID,
+      'is_test'=> $this->lastContribution['is_test'],
+      'is_pay_later' => 1,
+      'tax_amount'=> $this->lastContribution['tax_amount'],
+      'source' => $this->lastContribution['source'],
+      'campaign_id' => $this->lastContribution['campaign_id'],
+      'skipRecentView' => 1,
+      'skipLineItem' => 0, // TODO : Not sure of this flag, need to be tested
+      'membership_id'=> $this->currentMembershipID, // TODO : Not sure of this flag, need to be tested
+      'invoice_id' => md5(uniqid(rand(), TRUE))
+    ];
+
+    return civicrm_api3('Contribution', 'create', $newContributionParams)['id'];
+  }
+
+  /**
+   * Creates the membership offline auto-renew
+   * payment record.
+   *
+   * @param $contributionID
+   */
+  private function createMembershipPayment($contributionID) {
+    civicrm_api3('MembershipPayment', 'create', [
+      'membership_id' => $this->currentMembershipID,
+      'contribution_id' => $contributionID,
+    ]);
+  }
+
+  /**
+   * Creates line items for the membership
+   * contribution to be auto-renewed.
+   *
+   * @param $contributionID
+   */
+  private function createLineItem($contributionID) {
+    $lastContributionLineItem = civicrm_api3('LineItem', 'get', [
+      'sequential' => 1,
+      'return' => ['id', 'qty', 'unit_price', 'line_total', 'non_deductible_amount', 'tax_amount'],
+      'contribution_id' => $this->lastContribution['id'],
+      'options' => ['limit' => 1, 'sort' => 'id DESC'],
+    ])['values'][0];
+
+    $lineItemID = civicrm_api3('LineItem', 'create', [
+      'sequential' => 1,
+      'entity_table' => 'civicrm_membership',
+      'entity_id' => $this->currentMembershipID,
+      'contribution_id' => $contributionID,
+      //'label' => '' TODO : Student (33.33%)March 19th, 2018,
+      'qty' => $lastContributionLineItem['qty'],
+      'unit_price' => $lastContributionLineItem['unit_price'],
+      'line_total' => $lastContributionLineItem['line_total'],
+      'financial_type_id' => $this->financialTypesIDMap[$lastContributionLineItem['financial_type_id']],
+      'non_deductible_amount' => $lastContributionLineItem['non_deductible_amount'],
+      'tax_amount' => $lastContributionLineItem['tax_amount'],
+    ])['id'];
+
+
+    $lastContributionFinancialItem = civicrm_api3('FinancialItem', 'get', [
+      'sequential' => 1,
+      'return' => ['id', 'amount', 'currency'],
+      'entity_id' => $lastContributionLineItem['id'],
+      'entity_table' => 'civicrm_line_item',
+      'options' => ['limit' => 1, 'sort' => 'id DESC'],
+    ])['values'][0];
+
+    civicrm_api3('FinancialItem', 'create', [
+      'contact_id' => $this->contactID,
+      //'description' => '' TODO : Student (33.33%)March 19th, 2018,
```

Comment:
Again, this should be resolved before merging, right?

Reply:
done